[DT-M3] Create dt-maarifa API to receive contact and interaction data

adventureit opened the pull request:
The Maarifa RS currently uses built-in DT APIs to create contacts and comments. This mostly suits our needs right now, but some options to expand the Maarifa Plugin features could use their own API. 

Creating 3 endpoints just for the Maarifa Plugin that the RS will send data to directly and DT can handle some special cases better.

-POST /wp-json/dt-maarifa/v1/contacts
-POST /wp-json/dt-maarifa/v1/contacts/:id
-POST /wp-json/dt-maarifa/v1/contacts/:id/interactions

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

3 endpoints created.
1st - Create new contact/update if maarifa_data already exists - Ok, if using DT json formating;
2nd - Update based on DT id - Ok, if using DT json formating;
3rd - Create/update interactions - In progress. Create and update are Ok, if using DT json formating, but needs to add some fields.

It also remains to map the DT x Maarifa fields and use it before performing the above actions.

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data - Maping Maarifa x DT fields

Creating function mapFieldsToContact to map Maarifa x DT Fields and contemplate differences and exceptions.

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Mapping Maarifa x DT fields and correcting adding and updating comments.

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Corrections on Mapping Maarifa x DT fields (tags and milestones).

adventureit committed:
Merge branch 'master' of https://github.com/adventureit/dt-maarifa

cairocoder01 commented:
@adventureit Nice work getting this started. I tested the first endpoint and found a number of problems, so I didn't test the others yet since they will probably have the same ones. Once the bugs below are fixed, I can test the other 2 endpoints and look at the code in more detail.

## Simple Contact 
When sending to `POST /wp-json/dt-maarifa/v1/contacts`

```
{
    "id": 12345,
    "name": "John Doe",
    "email": "john.doe@test.com",
    "phone": "+1-9995551234",
    "street": "South Pole",
    "country": "antarctica",
    "gender": "male",
    "age": "35-44",
    "notes": "General info about contact"
}
```

I get the error:
```
{
    "code": "create_post",
    "message": "'notes' field expected to be an array",
    "data": null
}
```

When sending to `POST /wp-json/dt-maarifa/v1/contacts` 
with simple data above (without notes property)

I get the error:
```
{
    "code": "update_multi_select_fields",
    "message": "missing values field on: milestones",
    "data": {
        "status": 400
    }
}
```

However, it does create a contact and will succeed on the second try.

After the second try, I noticed these bugs:

1. [ ] the contact does not have the street/country values in the Locations or Address fields
2. [ ] the maarifa_data field is set to the id but should be the whole contact json object (this will cause the Maarifa tile on the contact screen to show the id and other info)
3. [ ] the contact should have a source of 'maarifa' added
4. [ ] the age field is not being set

## Full
When sending to `POST /wp-json/dt-maarifa/v1/contacts`

```
{
    "id": 45235,
    "name": "Jon Doe \/ Bob Johnson",
    "email": "[\"bobscott.explorer@gmail.com\",\"test@gmail.com\"]",
    "phone": "[\"+44-1234567890\",\"+1-9995551235\"]",
    "phonecode": "+44",
    "skype": "",
    "facebook": "",
    "street": "South Pole",
    "country": "antarctica",
    "ip_city": null,
    "ip_city_id": null,
    "gender": "male",
    "age": "45+",
    "background": "christian",
    "spiritual": "seeker",
    "tags": [{
        "id": 11,
        "name": "\u0623\u062d\u0644\u0627\u0645 \u0631\u0648\u062d\u064a\u0629",
        "alias": "spiritual-dreams"
    }, {
        "id": 19,
        "name": "\u0627\u0628\u0646 \u0627\u0644\u0633\u0644\u0627\u0645",
        "alias": "person-of-peace"
    }],
    "notes": "testest",
    "created": 1602661663,
    "last_updated": 1710502175,
    "last_seen": 1710502175,
    "first_contact": "whatsapp",
    "first_contact_details": null,
    "first_source": "-",
    "ds_user_id": "",
    "external_id": "20",
    "external_url": "https:\/\/maarifa.journey.tools\/demo",
    "location_details": null,
    "milestones": {
        "has bible": true,
        "studying": true,
        "profession": false,
        "can share": false,
        "sharing": false,
        "baptized": false,
        "baptizing": false,
        "in group": false,
        "starting groups": false
    }
}
```

I am noticing these bugs:

1. [ ] phone numbers are not split
2. [ ] emails are not split
3. [ ] location not set
4. [ ] age not set
5. [ ] Missing source of Maarifa
6. [ ] `maarifa_data` set to id instead of full object
7. [ ] Notes fails with message about needing to be an array. Removed that property to test the above.

## Code Formatting
You might notice that on this pull request, there is a notice that checks failed. If you click into that, you should be able to see that the PHPCS (PHP code sniffer) - which makes sure the code is formatted in a consistent way - is failing. You can run that check locally by using these commands on your command line:
```
composer install
composer run lint
composer run lintfix
```

The third one will try to automatically fix whatever it's able to. Mostly it fixes spacing of the code. The second will tell you exactly what other issues there are. Let me know if any of them are hard to understand.

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Correcting bugs

adventureit committed:
phpcs

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Fixing bugs:
-maarifa_data field
-age field
-phone numbers are not split
-emails are not split
-age not set
-maarifa_data set to id instead of full object
-Notes fails
-Code Formatting

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Fixing location.

cairocoder01 committed:
fix maarifa source

adventureit committed:
Merge branch 'master' of https://github.com/adventureit/dt-maarifa

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Adding source = maarifa

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Adding source = maarifa

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+            $fields_map['gender'] = $contact_map['gender'];
+        }
+
+        if ( !empty( $contact_map['street'] ) ) {
+
+            dt_write_log( 'Valor street' );
+            dt_write_log( $contact_map['street'] );
+
+            $fields_map['contact_address'] = [ [ 'value' => $contact_map['country'] ] ];
+        }
```
This if statement uses `street` in the condition, but then uses `country` to set the address. I think you want country in the condition so that they match.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+
+        }
+
+        if ( !empty( $contact_map['tags'] ) )
+            {
+
+            $int_count0 = 0;
+
+            foreach ( $contact_map['tags'] as $key => $value ) {
+
+                $fields_map['tags']['values'][$int_count0]['value'] = $contact_map['tags'][$key]['alias'];
+
+                $int_count0++;
+
+            }
+        }
```
This works, but there's two things that could clean it up:

- I don't think you need the int_count0 variable. You can just assign to `$fields_map['tags']['values'][] = [ 'value' => $value['alias']`. That `[]` means to append a new item to the array.
- In that code above, you'll see that I used `$value` instead of `$contact_map['tags'][$key]` since it should be the same thing from how you wrote the foreach statement. You get the key and the value, so you can just directly use the value.

cairocoder01 replied:
It looks like this suggestion hasn't been changed yet.

adventureit replied:
Something happened in my last commit and I lost the changes. I redid tit and just made a new commit.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+
+        if ( !empty( $contact_map['milestones'] ) ) {
+
+            $int_count = 0;
```
See comment above about not needing the int_count. You can just write `[]` to append a new item to an array.

cairocoder01 replied:
It looks like this suggestion hasn't been changed yet.

adventureit replied:
Something happened in my last commit and I lost the changes. I redid tit and just made a new commit.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+    }
+
+
+    public function update_post( WP_REST_Request $request ){
+
+        $url_params = $request->get_url_params();
+        $id_upd = $url_params['id'];
+
+        $fields_orig     = $request->get_json_params() ?? $request->get_body_params();
+
+        //Converts Maarifa field names to DT field names
+        $fields3 = $this->map_fields_to_contact( $fields_orig );
+
+
+        $this->update_maarifa_contact( $request, $id_upd, $fields3 );
+    }
```
Right now, this doesn't return anything in the JSON body of the response. The create endpoint returns the post, so this should return the same thing.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+                    $args['comment_meta'] = $value['meta'];
+                }
+
+                if ( isset( $value['id'] ) )
+                {//If comment_id exists, update the comment
+
+                    $id = $value['id'];
+
+                    dt_write_log( 'Add_interactions UPDATE' );
+                    dt_write_log( $id );
+
+                    $result = DT_Posts::update_post_comment( $id, $comment, true, $type, $args );
```
The `id` that is passed in the JSON data is not the id of the comment in DT. It is the ID of the interaction in the RS. So we can't use that to update an existing comment. We could probably use that id if we save it as extra meta data on the comment, but that is a little more complicated and we can handle that scenario later. For now, let's just always add new comments.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,672 @@
+                else { //If doesn't, create a new comment
+                    dt_write_log( 'Add_interactions CREATE' );
+
+                    $result = DT_Posts::add_post_comment( $url_params['post_type'], $url_params['id'], $comment, $type, $args, true, $silent );
```
I'm getting a date format error trying to save a comment/interaction. I'm guessing you haven't tested this yet, but just in case, here is the data I sent and response I received:

<img width="697" alt="image" src="https://github.com/user-attachments/assets/d8b82e6e-c1be-4169-b647-30bb225d07a2" />

cairocoder01 reviewed (changes requested):
@adventureit This is working much better now! Nicely done! Adding a contact seems to work great and updating contacts also seems to work well. I wasn't able to add an interaction/comment, but I'm guessing you haven't worked much on that part yet.

I left a few code comments of some things that can be cleaned and a few bugs.

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data

Fixing bugs in location (country), interactions, update (returning post), and adding comment.

cairocoder01 commented on rest-api/rest-api.php:
```diff
@@ -0,0 +1,637 @@
+        }
+        else //No milestones
+            {
+            $fields_map['milestones']['values'][1]['value'] = 0;
```
What is the intention with this code? I noticed that if I didn't send the `milestones` property in the API request, the activity comment in DT said "0 added to Faith Milestones".

cairocoder01 reviewed (changes requested):
Almost there. I just noticed a bug that if you don't send a milestones property in the API body when doing an update, you get an activity message in the DT UI that says "0 added to Faith Milestones"

cairocoder01 committed:
remove unused file

cairocoder01 committed:
Merge branch 'master' into adventureit-master

adventureit committed:
[DT-M3] Create dt-maarifa API to receive contact and interaction data - removing int_count

Removing int_count

adventureit commented:
Thank you. But I still have the  "0 added to Faith Milestones" bug to fix and I saw that every time I update a contact, the same location is being added multiple times. Working on it.

cairocoder01 commented:
Oh, sorry. I had tested the "0 added to Faith Milestones" and thought it was fixed. You'll have to just create another PR when you make those other changes since this one is merged and closed.
